src/Raven.Client/Documents/Session/Operations/BatchOperation.cs

```diff
@@ -77,21 +80,154 @@ CommandType GetCommandType(BlittableJsonReaderObject batchResult)
-                }
+                attachments.Add(attachment);
+                break;
+            }
+
+            documentInfo.MetadataInstance = null;
+            documentInfo.Metadata = _session.Context.ReadObject(documentInfo.Metadata, id);
+        }
+
+        private void HandleAttachmentPut(BlittableJsonReaderObject batchResult)
+        {
+            HandleAttachmentPutInternal(batchResult, CommandType.AttachmentPUT, nameof(PutAttachmentCommandData.Id), nameof(PutAttachmentCommandData.Name));
+        }
+
+        private void HandleAttachmentPutInternal(BlittableJsonReaderObject batchResult, CommandType type, string idFieldName, string attachmentNameFieldName)
+        {
+            var id = GetStringField(batchResult, type, idFieldName);
+
+            if (_session.DocumentsById.TryGetValue(id, out var documentInfo) == false)
+                return;
+
+            documentInfo.Metadata.Modifications = null;
+            documentInfo.Metadata.Modifications = new DynamicJsonValue(documentInfo.Metadata);
+
+            var attachments = documentInfo.Metadata.TryGet(Constants.Documents.Metadata.Attachments, out BlittableJsonReaderArray attachmentsJson)
+                ? new DynamicJsonArray(attachmentsJson)
+                : new DynamicJsonArray();
+
+            documentInfo.Metadata.Modifications[Constants.Documents.Metadata.Attachments] = attachments;
+
+            attachments.Add(new DynamicJsonValue
+            {
+                [nameof(AttachmentDetails.ChangeVector)] = GetStringField(batchResult, type, nameof(AttachmentDetails.ChangeVector)),
+                [nameof(AttachmentDetails.ContentType)] = GetStringField(batchResult, type, nameof(AttachmentDetails.ContentType)),
+                [nameof(AttachmentDetails.Hash)] = GetStringField(batchResult, type, nameof(AttachmentDetails.Hash)),
+                [nameof(AttachmentDetails.Name)] = GetStringField(batchResult, type, attachmentNameFieldName),
+                [nameof(AttachmentDetails.Size)] = GetLongField(batchResult, type, nameof(AttachmentDetails.Size))
+            });
+
+            documentInfo.MetadataInstance = null;
+            documentInfo.Metadata = _session.Context.ReadObject(documentInfo.Metadata, id);
+        }
+
+        private void HandlePatch(BlittableJsonReaderObject batchResult)
+        {
+            if (batchResult.TryGet(nameof(PatchStatus), out string statusAsString) == false)
+                ThrowMissingField(CommandType.PATCH, nameof(PatchStatus));
+
+            if (Enum.TryParse(statusAsString, ignoreCase: true, out PatchStatus status) == false)
+                ThrowMissingField(CommandType.PATCH, nameof(PatchStatus));
+
+            switch (status)
+            {
+                case PatchStatus.Created:
+                case PatchStatus.Patched:
+                    HandleDeleteInternal(batchResult, CommandType.PATCH); // deleting because it is impossible to know what was in the patch
```

Comment:
@ayende @maximburyak not sure what to do here. What do you think?

Reply:
It makes sense, don't think we should do anything else here

Reply:
This will mean that loading the document again will _return null_, no?
In this case, probably should evict it.

Note that I'm pretty sure that we already check that the document isn't already in the session / modified.

Reply:
This will remove it from the session (evict). So next 'Load' will issue a new request to the server to get it.

Downside of this is that following code fill not work:
```csharp
using (session)
{
    var x = load
    patch(x)
    saveChanges

    x.Name = 'Name' // not tracked anymore
    saveChanges // nothing will happen
}
```

Reply:
_should_ such code work?
That is a very strange thing to do, I think.
I also think we check if there are modifications to a patched entity.
I'm fine with keeping that check across `SaveChanges` calls.

Maybe just mark this as patched, meaning that unless user explicitly evict, we can't user the existing instance?

Reply:
What if... just before sending the DeferredPatchCommand... we will check if entity exists on the session. And if it does, then issue this command with additional parameter to return the document after patch in the 'replay'. We can then update the session properly?

Reply:
Done